pymeasure/instruments/hp/hp8753e.py

```diff
@@ -0,0 +1,394 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2023 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+__version__ = "0.1.1"
+
+import re
+import warnings
+from functools import cached_property
+from io import BytesIO
+from time import sleep
+from time import time as now
+
+import numpy as np
+from pyvisa import VisaIOError
+
+from pymeasure.adapters import PrologixAdapter
+from pymeasure.instruments import Instrument, RangeException, discreteTruncate
+
+
+class HP8753E(Instrument):
+    """Represents the HP8753E Vector Network Analyzer
+    and provides a high-level interface for taking scans of the
+    scattering parameters.
+
+    Template copied from Agilent 8722 class in pymeasure
```

Comment:
What do you want to tell with this line?

If the code is quite similar to the Agilent8722, could you inherit from the other one and just implement the changes?

Reply:
I could look to see if the commands are compatible or even close.

I think I need to import hplegacyinstrument at least as well.

Reply:
Honestly the Agilent8722 class could have been imported but all of the exceptions would need to be rewritten to reflect the proper VNA model being controlled.

Reply:
What exceptions would have to be changed?
Wouldn't that be a lot less work now (creating the new driver and tests) and later on (maintaining one instead of two drivers)?

Reply:
You have put a lot of limits in your code.
How confident are you, that your code (and the limits) apply to that other driver as well?

One approach would be to compare the settings in both manuals and decide on that base to use your driver for both instruments (maintaining the old, deprecated code in the agilent, though, for backward compatibility). 

I don't know whether you have the time for that work.

The backup solution is to add some line in each docstring, stating that the other device is quite similar.
That way someone, who knows the agilent could check whether your driver fits to the agilent and update the agilent one.
If you choose this way, you should probably open an issue as well.

Reply:
Actually I first tried to set those limits on initialization but it would error out. I now realize I would need to redefine those properties in the `__init__` function or initialize them to default values and override them in the `__init__` function if not using the default values.

I want the doc strings and exceptions to spell out the limits for what gets defined when the class is run and not the hard-coded values I currently have written. I know for a fact that both the 8722 and 8753 have a plethora of frequency ranges they can cover based on options purchased for the instrument.

I would prefer if there was a GPIB command to ask the instrument what the Min/Max value limits are but I don't know if a command exists for the 8722 or 8753 to do that like more modern instruments can with various SCPI commands. I might be able to program the limits based on the options reported using a lookup table instead.

To start, I'll just work on allowing for configuring the limits to all of the properties I have limits defined.

Since I don't have a 8722 to test against, I would rather not depreciate that code since it probably works for those that do and I can't ensure my code would too. If I can figure out how to make the Bytesio function the 8722ES class uses in `data_complex` work with my Prologix I would probably be a little less hesitant.

Reply:
Ok, let us not mix those two instruments (for this PR), so this PR touches only your new instrument.

> Actually I first tried to set those limits on initialization but it would error out. I now realize I would need to redefine those properties in the __init__ function or initialize them to default values and override them in the __init__ function if not using the default values.

For setting values later on, you could use `dynamic=True` in the properties, see [dynamic-properties documentation](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/properties.html#dynamic-properties).

> I want the doc strings and exceptions to spell out the limits for what gets defined when the class is run and not the hard-coded values I currently have written. I know for a fact that both the 8722 and 8753 have a plethora of frequency ranges they can cover based on options purchased for the instrument.

Our properties do not allow to change the documentation dynamically (see aforementioned documentation), however, if you use a `strict_range` validator, the error message will return the currently active limits.
In fact, in most cases a `strict` validator is better than a `truncated` one, as the latter does not give a feedback about the truncation, such that the user might erroneously think, that the value is valid.

> Since I don't have a 8722 to test against, I would rather not depreciate that code since it probably works for those that do and I can't ensure my code would too.

I did not mean to deprecate the whole 8722, but keep the methods of the 8722, which are already deprecated. As we do not combine those two instruments (at the moment). Don't worry about it anymore.

Reply:
I added dynamic properties where appropriate. Otherwise I think everything in this conversation has been resolved.